src/core/detail/btree_internal.h

```diff
@@ -0,0 +1,458 @@
+
+  // Rebalance a full child at position pos, at which we tried to insert at insert_pos.
+  // Returns the node and the position to insert into if rebalancing successful.
+  // Returns nullptr if rebalancing did not succeed.
+  std::pair<BTreeNode*, unsigned> RebalanceChild(unsigned pos, unsigned insert_pos);
+
+  // Inserts item into a leaf node.
+  // Assumes: the node is IsLeaf() and has some space.
+  void LeafInsert(unsigned index, Key_t item) {
+    assert(IsLeaf() && NumItems() < MaxItems());
+    InsertItem(index, item);
+  }
+
+  void InnerInsert(unsigned index, Key_t item, BTreeNode* child) {
+    InsertItem(index, item);
+    SetChild(index + 1, child);
+  }
+
+  // Tries to merge the child at position pos with its sibling.
+  // If we did not succeed to merge, we try to rebalance.
+  // Returns retired BTreeNode* if children got merged and this parent node's children
+  // count decreased, otherwise, we return nullptr (rebalanced).
+  BTreeNode* MergeOrRebalanceChild(unsigned pos);
+
+  void Validate(Key_t upper_bound) const;
+
+ private:
+  void RebalanceChildToLeft(unsigned child_pos, unsigned count);
+  void RebalanceChildToRight(unsigned child_pos, unsigned count);
+
+  void MergeFromRight(Key_t key, BTreeNode* right);
+
+  void InsertItem(unsigned index, Key_t item) {
+    assert(index <= num_items_);
+    assert(index == 0 || Key(index - 1) < item);
+    assert(index == num_items_ || Key(index) > item);
+
+    ShiftRight(index);
+    SetKey(index, item);
+  }
+
+  uint8_t* KeyPtr(unsigned index) {
+    uint8_t* ptr = reinterpret_cast<uint8_t*>(this) + kKeyOffset + kKeySize * index;
+    return ptr;
+  }
+
+  const uint8_t* KeyPtr(unsigned index) const {
+    const uint8_t* ptr = reinterpret_cast<const uint8_t*>(this) + kKeyOffset + kKeySize * index;
+    return ptr;
+  }
+
+  union {
+    struct {
+      uint64_t num_items_ : 7;
+      uint64_t leaf_ : 1;
+      uint64_t : 56;
+    };
+
+    uint64_t val_;
+  };
+
+  static_assert(kMaxLeafKeys < 128, "num_items_ is 7 bits");
+};
+
+// Contains parent/index pairs. Meaning that node0->Child(index0) == node1.
+template <typename Policy> class BTreePath {
+  static constexpr unsigned kMaxDepth = 16;
+
+ public:
+  void Push(BTreeNode<Policy>* node, unsigned pos) {
+    assert(depth_ < kMaxDepth);
+    record_[depth_].node = node;
+    record_[depth_].pos = pos;
+    depth_++;
+  }
+
+  unsigned Depth() const {
+    return depth_;
+  }
+
+  std::pair<BTreeNode<Policy>*, unsigned> Last() const {
+    return {record_[depth_ - 1].node, record_[depth_ - 1].pos};
+  }
+
+  BTreeNode<Policy>* Node(unsigned i) const {
+    assert(i < depth_);
+    return record_[i].node;
+  }
+
+  unsigned Position(unsigned i) const {
+    assert(i < depth_);
+    return record_[i].pos;
+  }
+
+  void Pop() {
+    assert(depth_ > 0u);
+    depth_--;
+  }
+
+ private:
+  struct Record {
+    BTreeNode<Policy>* node;
+    unsigned pos;
+  };
+
+  Record record_[kMaxDepth];
```

Comment:
Use `std::array<Record, kMaxDepth>` instead?

Reply:
I change it but inside I weakly disagree. I feel that array is c++ struct that has advantage in SOME cases but in those it does not, it's just another include.